vnext/Microsoft.ReactNative/Fabric/Composition/TextInput/WindowsTextInputComponentView.cpp

```diff
@@ -772,6 +776,83 @@ void WindowsTextInputComponentView::UpdateText(const std::string &str) noexcept
       m_textServices->TxSendMessage(EM_SETTYPOGRAPHYOPTIONS, 0x1000 | 0x2000, 0x1000 | 0x2000, nullptr));
 }
 
+void WindowsTextInputComponentView::DrawPlaceholderText() noexcept {
+  m_needsRedraw = true;
+  if (m_cDrawBlock) {
+    return;
+  }
+
+  if (!m_drawingSurface)
+    return;
+
+  // Begin our update of the surface pixels. If this is our first update, we are required
+  // to specify the entire surface, which nullptr is shorthand for (but, as it works out,
+  // any time we make an update we touch the entire surface, so we always pass nullptr).
+  winrt::com_ptr<ID2D1DeviceContext> d2dDeviceContext;
+  POINT offset;
+
+  winrt::com_ptr<Composition::ICompositionDrawingSurfaceInterop> drawingSurfaceInterop;
+  m_drawingSurface.as(drawingSurfaceInterop);
+
+  m_drawing = true;
+  if (CheckForDeviceRemoved(drawingSurfaceInterop->BeginDraw(d2dDeviceContext.put(), &offset))) {
+    d2dDeviceContext->Clear(D2D1::ColorF(D2D1::ColorF::Black, 0.0f));
+    assert(d2dDeviceContext->GetUnitMode() == D2D1_UNIT_MODE_DIPS);
+    const auto dpi = m_layoutMetrics.pointScaleFactor * 96.0f;
+    float oldDpiX, oldDpiY;
+    d2dDeviceContext->GetDpi(&oldDpiX, &oldDpiY);
+    d2dDeviceContext->SetDpi(dpi, dpi);
+
+    facebook::react::LayoutConstraints constraints;
+    constraints.maximumSize.width = static_cast<FLOAT>(offset.x) + static_cast<FLOAT>(m_imgWidth);
+    constraints.maximumSize.height = static_cast<FLOAT>(offset.y) + static_cast<FLOAT>(m_imgHeight);
+
+    // set brush color
+    winrt::com_ptr<ID2D1SolidColorBrush> brush;
+    if (m_placeholderTextColor) {
+      auto color = m_placeholderTextColor.AsD2DColor();
+      winrt::check_hresult(d2dDeviceContext->CreateSolidColorBrush(color, brush.put()));
+    } else {
+      winrt::check_hresult(
+          d2dDeviceContext->CreateSolidColorBrush(D2D1::ColorF(D2D1::ColorF::Gray, 1.0f), brush.put()));
+    }
+
+    // Create a fragment with text attributes
+    winrt::com_ptr<::IDWriteTextLayout> textLayout = nullptr;
+    facebook::react::AttributedString attributedString;
+    facebook::react::AttributedString::Fragment fragment1;
+    facebook::react::TextAttributes textAttributes = m_props->textAttributes;
+    if (std::isnan(m_props->textAttributes.fontSize)) {
+      textAttributes.fontSize = 12.0f;
+    }
+    fragment1.string = m_placeholderText;
+    fragment1.textAttributes = textAttributes;
+    attributedString.appendFragment(fragment1);
+
+    facebook::react::TextLayoutManager::GetTextLayout(
+        facebook::react::AttributedStringBox(attributedString), {} /*TODO*/, constraints, textLayout);
+
+    // draw text
+    d2dDeviceContext->DrawTextLayout(
+        D2D1::Point2F(
```

Comment:
Can we make this part a separate function, and call it from within DrawText, if the text is empty?
That way we would reduce the duplicated DrawBlock, and begin draw etc logic, and hopefully it would already call without needing the extra DrawText call in OnTextUpdated.

Reply:
Oh that's such a good idea! That seemed to have made the logic a little more easier to follow too, thanks! :)